utils/trace_parser.py

```diff
@@ -0,0 +1,182 @@
+                events.append(Event(name, start, end, optype))
+            elif evtype == "B":
+                partialEvents[name] = Event(name, start, end, optype)
+            elif evtype == "E":
+                if not name in partialEvents:
+                    # This is a bug in Glow tracing, but ignore for now.
+                    continue
+                ev = partialEvents[name]
+                ev.end = start
+                events.append(ev)
+
+    return events
+
+# Find all enclosed events and move them to be children.
+
+
+def stackEvents(events):
+    events = sorted(events, key=attrgetter("end"), reverse=True)
+    events = sorted(events, key=attrgetter("start"))
+    result = []
+    lastEvent = None
+    for ev in events:
+        if lastEvent:
+            if lastEvent.totalOverlap(ev):
+                lastEvent.addChild(ev)
+                continue
+            if lastEvent.children:
+                lastEvent.children = stackEvents(lastEvent.children)
+                lastEvent.updateChildTime()
+        lastEvent = ev
+        result.append(ev)
+    if lastEvent.children:
+        lastEvent.children = stackEvents(lastEvent.children)
+        lastEvent.updateChildTime()
+    return result
+
+# Accumulate Event durations by a key produced by keyfunc. Keyfunc is a lambda which takes an Event as a parameter.
+
+
+def dumpAccumulate(events, keyfunc):
+    nameMap = defaultdict(list)
+    for ev in events:
+        name = keyfunc(ev)
+        nameMap[name].append(ev.selfTime())
+
+    layers = []
+    for (name, times) in nameMap.items():
+        layers.append((name, len(times), numpy.mean(times),
+                       numpy.std(times), numpy.sum(times)))
+
+    # Iterate sorted by total time.
+    for (name, num, mean, stddev, total) in sorted(layers, key=itemgetter(4), reverse=True):
+        print("%s %s events, mean: %s, stddev: %s, total: %s" % (
+            name, num, formatUs(mean), formatUs(stddev), formatUs(total)))
+    print()
+    print()
+
+
+parser = argparse.ArgumentParser(description='process trace json')
+parser.add_argument('filename', type=str,
+                    help='filename for trace file to load')
+parser.add_argument("--layers", action='store_true')
+parser.add_argument("--kinds", action='store_true')
+
+args = parser.parse_args()
+events = loadEvents(args.filename)
+
+# Stack events so we can determine selfTime.
+stacked = stackEvents(events)
+
+if args.layers:
+    dumpAccumulate(events, lambda ev: "%s (%s)" % (ev.name, ev.optype))
+
+if args.kinds:
+    dumpAccumulate(events, lambda ev: ev.optype)
+
+# Ensure events are sorted by startTime.
+stacked = sorted(stacked, key=attrgetter("start"))
+totalTime = stacked[-1].end - stacked[0].start
+coveredTime = 0
+for ev in stacked:
+    coveredTime += ev.end - ev.start
+
+print("Total time of trace:", formatUs(totalTime))
+print("Time covered by events:", formatUs(coveredTime))
+print("Unattributed time:", formatUs(totalTime-coveredTime))
```

Comment:
Let's wrap this in a `main()` function and call it like
```
if __name__ == "__main__":
    main()
```

in case we need to reuse any of the classes or modules defined in this file elsewhere.